toolkit/tools/pkg/imagecustomizerlib/partitionutils.go

```diff
@@ -791,3 +808,153 @@ func getPartitionNum(partitionLoopDevice string) (int, error) {
+	if err != nil {
+		return nil, fmt.Errorf("failed to list BTRFS subvolumes: %w\nstderr: %s", err, stderr)
+	}
+
+	var subvolumes []string
+	lines := strings.Split(strings.TrimSpace(stdout), "\n")
+
+	for _, line := range lines {
+		if strings.TrimSpace(line) == "" {
+			continue
+		}
+
+		// Parse btrfs subvolume list output format:
+		// ID 256 gen 7 top level 5 path @
+		// ID 257 gen 7 top level 5 path @home
+		parts := strings.Fields(line)
+		if len(parts) >= 2 && parts[len(parts)-2] == "path" {
+			subvolPath := parts[len(parts)-1]
+			// Skip special subvolumes that start with <
+			if !strings.HasPrefix(subvolPath, "<") {
+				subvolumes = append(subvolumes, subvolPath)
+			}
+		}
+	}
+
+	logger.Log.Debugf("Found BTRFS subvolumes: %v", subvolumes)
+	return subvolumes, nil
+}
+
+// mountBtrfsSubvolume mounts a specific BTRFS subvolume
+func mountBtrfsSubvolume(devicePath, subvolume, mountPoint string) (*safemount.Mount, error) {
+	mountOptions := fmt.Sprintf("subvol=%s,ro", subvolume)
+	mount, err := safemount.NewMount(devicePath, mountPoint, "btrfs", unix.MS_RDONLY, mountOptions, true)
+	if err != nil {
+		return nil, fmt.Errorf("failed to mount BTRFS subvolume %s:\n%w", subvolume, err)
+	}
+	return mount, nil
+}
+
+// processBtrfsSubvolumes is a helper function that handles the common pattern of mounting BTRFS,
+// listing subvolumes, and iterating through them with a callback function
+func processBtrfsSubvolumes(diskPartition *diskutils.PartitionInfo, tmpDir string,
+	callback func(subvolume string, subvolMount *safemount.Mount) (bool, error),
+) (bool, error) {
+	// First mount the BTRFS filesystem to discover subvolumes
+	partitionMount, err := safemount.NewMount(diskPartition.Path, tmpDir, diskPartition.FileSystemType,
+		unix.MS_RDONLY, "", true)
+	if err != nil {
+		return false, fmt.Errorf("failed to mount partition (%s):\n%w", diskPartition.Path, err)
+	}
+	defer partitionMount.Close()
+
+	// List actual subvolumes using btrfs tools
+	subvolumes, err := listBtrfsSubvolumes(tmpDir)
+	if err != nil {
+		logger.Log.Debugf("Failed to list BTRFS subvolumes for partition %s:\n%v", diskPartition.Path, err)
+		// Fallback to checking default location with callback
+		return callback("", partitionMount)
+	}
+
+	// Close the current mount and mount each subvolume individually
+	err = partitionMount.CleanClose()
+	if err != nil {
+		return false, fmt.Errorf("failed to close partition mount (%s):\n%w", diskPartition.Path, err)
+	}
+
+	// Check each subvolume
+	for _, subvolume := range subvolumes {
+		subvolMount, err := mountBtrfsSubvolume(diskPartition.Path, subvolume, tmpDir)
+		if err != nil {
+			logger.Log.Debugf("Failed to mount BTRFS subvolume %s:\n%v", subvolume, err)
+			continue
+		}
+
+		found, err := callback(subvolume, subvolMount)
+		subvolMount.Close() // Always close the subvolume mount
+
+		if err != nil {
+			logger.Log.Debugf("Error processing BTRFS subvolume %s:\n%v", subvolume, err)
```

Comment:
Ignoring the error here is silly and makes the `readFstabFromBtrfsPartition` function extremely confusing. If the error isn't a problem, then let the callback return `nil`.